module/VuFind/src/VuFind/Auth/AbstractShibboleth.php

```diff
@@ -0,0 +1,345 @@
+        // we also want to be careful not to replace a non-blank password with a
+        // blank one in case the auth mechanism fails to provide a password on
+        // an occasion after the user has manually stored one. (For discussion,
+        // see https://github.com/vufind-org/vufind/pull/612). Note that in the
+        // (unlikely) scenario that a password can actually change from non-blank
+        // to blank, additional work may need to be done here.
+        if (!empty($user->cat_username)) {
+            $user->saveCredentials(
+                $user->cat_username,
+                empty($catPassword) ? $user->getCatPassword() : $catPassword
+            );
+        }
+
+        $this->storeShibbolethSession($request);
+
+        // Save and return the user object:
+        $user->save();
+        return $user;
+    }
+
+    /**
+     * Return shibboleth configuration.
+     *
+     * @param Request $request Request object
+     *
+     * @return array
+     */
+    abstract protected function getShibbolethConfiguration($request);
+
+    /**
+     * Extract required user attributes from the configuration.
+     *
+     * @param array $config shibboleth configuration
+     *
+     * @return array      Only username and attribute-related values
+     * @throws AuthException
+     */
+    protected function getRequiredAttributes($config)
+    {
+        // Special case -- store username as-is to establish return array:
+        $sortedUserAttributes = [];
+
+        // Now extract user attribute values:
+        $shib = $this->getConfig()->Shibboleth;
+        foreach ($shib as $key => $value) {
+            if (preg_match("/userattribute_[0-9]{1,}/", $key)) {
+                $valueKey = 'userattribute_value_' . substr($key, 14);
+                $sortedUserAttributes[$value] = $shib['$valueKey'] ?? null;
+
+                // Throw an exception if attributes are missing/empty.
+                if (empty($sortedUserAttributes[$value])) {
+                    throw new AuthException(
+                        "User attribute value of " . $value . " is missing!"
+                    );
+                }
+            }
+        }
+
+        return $sortedUserAttributes;
+    }
+
+    /**
+     * Add session id mapping to external_session table for single logout support
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing
+     * account credentials.
+     *
+     * @return \VuFind\Db\Row\User Object representing logged-in user.
+     */
+    protected function storeShibbolethSession($request)
+    {
+        $shib = $this->getConfig()->Shibboleth;
+        if (!isset($shib['session_id'])) {
+            return;
+        }
+        $shibSessionId = getAttribute($request, $shib['session_id']);
+        if (null === $shibSessionId) {
+            return;
+        }
+        $localSessionId = $this->sessionManager->getId();
```

Comment:
The sessionManager property is used here in AbstractShibboleth, but it is only defined and populated in the regular Shibboleth class; I assume this is going to cause problems in ShibbolethWithWAYF. You might want to review this part of the refactoring.